docs/proposals/auditing.md

```diff
@@ -0,0 +1,260 @@
+
+### Events
+
+```go
+package audit
+
+type Event struct {
+  Level Level
+
+  // opening
+  ID string
+  Timestamp time.Timestamp
+  IP string
+  Method string
+  URI string
+  User string
+  AsUser string
+  Namespace, Name string
+  GroupKind unversioned.GroupKind
+
+  // unmodified request object before authn, auth & admission
+  RequestObject runtime.Unstructured
+
+  // storage object
+  OldObject runtime.Object
+  NewObject runtime.Object
+
+  // closing
+  Response string
+}
+```
+
+### Concurrency/Transaction Model for Events
+
+TODO: define (and maybe adapt interface) how synchronization of the writable audit event of a context takes place.
+
+### Policy
+
+```go
+package audit
+
+type Level string
+
+const (
+  // only logs http request level information without payload inspection
+  HttpHeadersLevel = "HttpHeaders"
+  // in addition to `HttpHeaders` the request payload is logged (if the output backend decides so)
+  RequestObjectLevel = "RequestObject"
+  // in addition to `RequestObject` the old and new object are intercepted on
+  // the storage layer and passed to the output backend. The output backend can // decide whether full objects are logged or a differential output is produced.
+  StorageObjectLevel = "StorageObject"
+)
+
+type Action interface {
+  Level() Level
+  Value(key string) string
+}
+
+type Policy interface {
+  func Action(kind, namespace, method, user string) Action
+}
+```
+
+TODO: how do the storage layer and the http handlers get access to the policy?
+
+### Policy Actions
+
+TODO: define what we need next to the audit level
+
+### Output Backend Interface
+
+```go
+package audit
+
+type OutputBackend interface {
+  Log(e *Event, a Action) error
+}
+```
+
+### Apiserver Command Line Flags
+
+```bash
+$ kube-apiserver --audit-output file:path=/var/log/apiserver-audit.log,rotate=1d,max=1024MB,format=json \
+  --audit-policy user=EndUser,kind=ReplicationController:output=diff,exclude=status
+```
+
+## Sensible (not necessarily sequential) Milestones of Implementation
+
+1. add `audit.Event` and `audit.OutputBackend` and implement https://github.com/kubernetes/kubernetes/pull/27087/'s basic auth using them
+2. add deep inspection on the storage level to the old and the new object
+3. add output level policy support (versus static policy in the initial implementation)
+
+## Examples
+
+### Audit Policy
+
+- never log `*.Status`
+- never log `Secret.Spec`
+- never log `Endpoint`s
+- never log if user is "kubelet" or "master"
+- only log headers for WATCH, GET
+- only log the request object on POST and PUT
+- never log OPTIONS
```

Comment:
That's the plan.

On Thu, Jul 28, 2016 at 2:54 PM, Maciej Szulik <notifications@github.com>
wrote:

> In docs/proposals/auditing.md
> <https://github.com/kubernetes/kubernetes/pull/29443#discussion_r72615813>
> :
>
> > +
> > +1. add `audit.Event` and `audit.OutputBackend` and implement https://github.com/kubernetes/kubernetes/pull/27087/'s basic auth using them
> > +2. add deep inspection on the storage level to the old and the new object
> > +3. add output level policy support (versus static policy in the initial implementation)
> > +
> > +## Examples
> > +
> > +### Audit Policy
> > +
> > +- never log `*.Status`
> > +- never log `Secret.Spec`
> > +- never log `Endpoint`s
> > +- never log if user is "kubelet" or "master"
> > +- only log headers for WATCH, GET
> > +- only log the request object on POST and PUT
> > +- never log OPTIONS
>
> Define some kind of DSL for that?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29443/files/bc2f0badd7ef42014757d9eb8625ed249c17f152#r72615813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAskC3H-74ypoOpzX3e-aIV-ySSiXmAzks5qaKaWgaJpZM4JSliE>
> .
>